File: src/_cpp/TowerDefenseBase/game/unit/Bullet.cpp
/****************************************************************************
Copyright (c) 2014-2015 Vladimir Tolmachev (Volodar)

This file is part of game Greeco Defense

Greeco Defense is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Greeco Defense is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Greeco Defense.  If not, see <http://www.gnu.org/licenses/>.
****************************************************************************/

#include "Bullet.h"
NS_CC_BEGIN


Bullet::Bullet()
: _steering( false )
, _isStuck(false)
, _trajectory( Trajectory::line )
{
	_parabolicParams.H = 0;
	_parabolicParams.timer = 0;
	_parabolicParams.duration = CCRANDOM_MINUS1_1() * 0.2f + 0.5f;
}

Bullet::~Bullet( )
{
}

bool Bullet::init( const std::string& path, Unit::Pointer base, Unit::Pointer target, float startAngle, Point startPosition )
{
	do
	{
		CC_BREAK_IF( !Unit::init( "", path ) );
		_base = base;
		_target = target;

		setDamageBySector( base->getDamageBySector() );
		setAllowTargets( base->getAllowTargets() );
		setType( UnitType::tower );

		auto part = _target->getParamCollection().get( _bopyPart, "" );
		auto rand = _target->getParamCollection().get( "random_bullet", "" );
		_targetPointOffset = strToPoint( part );
		if( rand.empty() == false )
		{
			auto point = strToPoint( rand );
			_targetPointOffset.x += CCRANDOM_MINUS1_1() * point.x / 2;
			_targetPointOffset.y += CCRANDOM_MINUS1_1() * point.y / 2;
		}

		_startPoint = startPosition;
		updateTargetPoint();
		prepare();

		setPosition( startPosition );
		setRotation( startAngle );
		float z = _unitLayer == UnitLayer::sky ? 9000 : -startPosition.y;
		setLocalZOrder( z + _additionalZorder );
		return true;
	}
	while( false );
	return false;
}

bool Bullet::setProperty( const std::string & stringproperty, const std::string & value )
{
	if( stringproperty == "trajectory" )
	{
		if( value == "line" )
			_trajectory = Trajectory::line;
		else if( value == "parabolic" )
			_trajectory = Trajectory::parabolic;
		else
			assert( 0 );
	}
	else if( stringproperty == "parabolicheight" )
		_parabolicParams.H = strToFloat( value );
	else if( stringproperty == "steering" )
		_steering = strToBool( value );
	else if( stringproperty == "bodypart" )
		_bopyPart = value;
	else if( stringproperty == "parabolic_duration" )
		_parabolicParams.duration = CCRANDOM_MINUS1_1() * 0.2f + strToFloat( value );
	else if( stringproperty == "stuck" )
		_isStuck = strToBool( value );
	else 
		return Unit::setProperty( stringproperty, value );
	return true;
}

void Bullet::update( float dt )
{
	Unit::update( dt );
}

void Bullet::clear()
{
	_base.reset( nullptr );
	_target.reset( nullptr );
}

void Bullet::readyfire_update( float dt )
{
	Unit::readyfire_update( dt );

	Point pos = computePosition( dt );
	turn( pos );
	setPosition( pos );
	float z = _unitLayer == UnitLayer::sky ? 9000 : -pos.y;
	setLocalZOrder( z + _additionalZorder );

}

void Bullet::on_die()
{
	if( _base && _target )
	{
		Point p0 = getPosition();
		Point p1 = _target->getPosition() + strToPoint( _target->getParamCollection().get( _bopyPart, "" ) );
		float distance = p0.getDistance( p1 );
		float radius = getRadius();

		getEffect().copyFrom( _base->getEffect() );
		bool applyDamage( false );
		applyDamage = applyDamage || (distance < radius);
		applyDamage = applyDamage || getDamageBySector();
		if( applyDamage )
		{
			setType( _base->getType() );
			applyDamageToTarget( _target );

			Unit::on_die();
			
			if( _isStuck )
			{
				auto newParent = _target->getChildByName( "bullet_node" );
				if( !newParent )newParent = _target;

				auto pos = getPosition();
				pos = convertToWorldSpace( Point::ZERO );
				pos = newParent->convertToNodeSpace( pos );
				setPosition( pos );
				retain();
				removeFromParent();
				newParent->addChild( this, -10 );
				release();
			}

			_base.reset( nullptr );
			_target.reset( nullptr );
		}
	}
}

void Bullet::turn( const Point& newPos )
{
	if( _target )
	{
		switch( _trajectory )
		{
			case Trajectory::line:
			{
				Point p0 = newPos;
				Point p1 = _target->getPosition();
				float angle = getDirectionByVector( p1 - p0 );
				setRotation( angle );
				break;
			}
			case Trajectory::parabolic:
			{
				Point p0 = getPosition();
				Point p1 = newPos;
				float angle = getDirectionByVector( p1 - p0 );
				setRotation( angle );
				break;
			}
			default:
				assert( 0 );
		}
	}
}

void Bullet::prepare()
{
	switch( _trajectory )
	{
		case Trajectory::line:
			break;
		case Trajectory::parabolic:
			if( _parabolicParams.H == 0 )
				_parabolicParams.H = 200;
			_parabolicParams.timer = 0;
			break;
	}
}

Point Bullet::updateTargetPoint()
{
	if( _targetPoint.equals( Point::ZERO ) || _steering )
	{
		_targetPoint = _target->getPosition() + _targetPointOffset;
	}
	return _targetPoint;
}

Point Bullet::computePosition( float dt )
{
	updateTargetPoint();

	switch( _trajectory )
	{
		case Trajectory::line:
			return computePositionLine( dt );
		case Trajectory::parabolic:
			return computePositionParabolic( dt );
		default:
			assert( 0 );
	}
	return getPosition();
}

Point Bullet::computePositionLine( float dt )
{
	Point p0 = getPosition();
	Point p1 = _targetPoint;
	float distance = p0.getDistance( p1 );
	float velocity = getMover().getVelocity();
	float t = dt*velocity;
	t = std::min( t, (p1 - p0).getLength() );
	Point p = p0 + (p1 - p0).getNormalized() * t;

	
	if( checkRadiusByEllipse( p, _targetPoint, getRadius() ) )
		push_event( event_die );

	return p;
}

Point Bullet::computePositionParabolic( float dt )
{
	/* t = [0;1] */
	auto parabolic = [this]( float t )
	{ return -(2 * t - 1)*(2 * t - 1) + 1; };
	auto line = [this]( float t )
	{ return _startPoint + (_targetPoint - _startPoint) * t; };

	_parabolicParams.timer += dt;
	float t = _parabolicParams.timer / _parabolicParams.duration; // 1 - duration;
	t = std::min( t, 1.f );
	Point pos = line( t );
	pos.y += parabolic( t ) * _parabolicParams.H;

	if( t >= 0.999f )
		push_event( event_die );

	return pos;
}


NS_CC_END

File: src/_eu/game/unit/Bullet.js
//Define namespace
var EU = EU || {};


EU.Bullet = EU.Unit.extend({
    /** For Test Instance of */
    __Bullet: true,
    Trajectory: {
        line: 0,
        parabolic: 1,
    },
    _bopyPart: null,
    _trajectory: null,
    /** @type {EU.Unit} */
    _base: null,
    /** @type {EU.Unit} */
    _target: null,
    /** @type {cc.p} */
    _startPoint: null,
    /** @type {cc.p} */
    _targetPoint: null,
    /** @type {cc.p} */
    _targetPointOffset: null,
    steering: null,
    _isStuck: null,
    _parabolicParams: {
        H: null,
        timer: null,
        duration: null,
    },
    onExit: function( )
    {
    },
    ctor: function( path, base, target, startAngle, startPosition )
    {
        //if( !Unit.init( "", path ) ) return false;
        //this._super("", path);
        EU.Unit.prototype.ctor.call(this,"", path);
        this.steering = false ;
        this._isStuck = false ;
        this._trajectory( this.Trajectory.line );
        this._parabolicParams.H = 0;
        this._parabolicParams.timer = 0;

        this._parabolicParams.duration = cc.randomMinus1To1() * 0.2 + 0.5;

        this._base = base;
        this._target = target;

        this.setDamageBySector( base.getDamageBySector() );
        this.setAllowTargets( base.getAllowTargets() );
        this.setType( EU.UnitType.tower );

        var part = this._target.getParamCollection().get( this._bopyPart, "" );
        var rand = this._target.getParamCollection().get( "random_bullet", "" );
        this._targetPointOffset = EU.Common.strToPoint( part );
        if( EU.xmlLoader.stringIsEmpty(rand) == false )
        {
            var point = EU.Common.strToPoint( rand );
            this._targetPointOffset.x += cc.randomMinus1To1() * point.x / 2;
            this._targetPointOffset.y += cc.randomMinus1To1() * point.y / 2;
        }

        this._startPoint = startPosition;
        this.updateTargetPoint();
        this.prepare();

        this.setPosition( startPosition );
        this.setRotation( startAngle );
        var z = this._unitLayer == EU.UnitLayer.sky ? 9000 : -startPosition.y;
        this.setLocalZOrder( z + this._additionalZorder );
        return true;
    },
    setProperty: function(stringproperty, value )
    {
        if( stringproperty == "trajectory" )
        {
            if( value == "line" )
                this._trajectory = this.Trajectory.line;
            else if( value == "parabolic" )
                this._trajectory = this.Trajectory.parabolic;
            else
                EU.assert( 0 );
        }
        else if( stringproperty == "parabolicheight" )
            this._parabolicParams.H = EU.Common.strToFloat( value );
        else if( stringproperty == "steering" )
            this.steering = EU.Common.strToBool( value );
        else if( stringproperty == "bodypart" )
            this._bopyPart = value;
        else if( stringproperty == "parabolic_duration" )
            this._parabolicParams.duration = cc.randomMinus1To1() * 0.2 + EU.Common.strToFloat( value );
        else if( stringproperty == "stuck" )
            this._isStuck = EU.Common.strToBool( value );
        else
            return Unit.setProperty( stringproperty, value );
        return true;
    },
    update: function( dt )
    {
        EU.Unit.prototype.update.call(this, dt );
    },
    clear: function()
    {
        this._base.reset( null);
        this._target.reset( null);
    },
    readyfire_update: function( dt )
    {
        EU.Unit.prototype.readyfire_update.call(this, dt );

        var pos = this.computePosition( dt );
        this.turn( pos );
        this.setPosition( pos );
        var z = this._unitLayer == EU.UnitLayer.sky ? 9000 : -pos.y;
        this.setLocalZOrder( z + this._additionalZorder );
    },
    on_die: function()
    {
        if( this._base && this._target )
        {
            var p0 = getPosition();
            var p1 = this._target.getPosition() + EU.Common.strToPoint( this._target.getParamCollection().get( this._bopyPart, "" ) );
            var distance = p0.getDistance( p1 );
            var radius = getRadius();
    
            this.getEffect().copyFrom( this._base.getEffect() );
            var applyDamage = false ;
            applyDamage = applyDamage || (distance < radius);
            applyDamage = applyDamage || this.getDamageBySector();
            if( applyDamage )
            {
                this.setType( this._base.getType() );
                this.applyDamageToTarget( this._target );

                EU.Unit.prototype.on_die.call(this, dt );

                if( this._isStuck )
                {
                    var newParent = this._target.getChildByName( "bullet_node" );
                    if( !newParent )newParent = this._target;
    
                    var pos = this.getPosition();
                    pos = this.convertToWorldSpace( cc.POINT_ZERO);
                    pos = newParent.convertToNodeSpace( pos );
                    this.setPosition( pos );
                    this.retain();
                    this.removeFromParentAndCleanup(true);
                    newParent.addChild( this, -10 );
                    release();
                }
    
                this._base.reset( null);
                this._target.reset( null);
            }
        }
    },
    turn: function( newPos )
    {
        if( this._target )
        {
            switch( this._trajectory )
            {
                case this.Trajectory.line:
                {
                    var p0 = newPos;
                    var p1 = this._target.getPosition();
                    var angle = this.getDirectionByVector( p1 - p0 );
                    this.setRotation( angle );
                    break;
                }
                case this.Trajectory.parabolic:
                {
                    var p0 = this.getPosition();
                    var p1 = newPos;
                    var angle = this.getDirectionByVector( p1 - p0 );
                    this.setRotation( angle );
                    break;
                }
                default:
                    EU.assert( 0 );
            }
        }
    },
    
    prepare: function()
    {
        switch( this._trajectory )
        {
            case this.Trajectory.line:
                break;
            case this.Trajectory.parabolic:
                if( this._parabolicParams.H == 0 )
                    this._parabolicParams.H = 200;
                this._parabolicParams.timer = 0;
                break;
        }
    },
    updateTargetPoint: function()
    {
        if( cc.pointEqualToPoint(this._targetPoint , cc.POINT_ZERO) || this.steering )
        {
            this._targetPoint = cc.pAdd(this._target.getPosition() , this._targetPointOffset);
        }
        return this._targetPoint;
    },
    computePosition: function(  dt )
    {
        this.updateTargetPoint();
    
        switch( this._trajectory )
        {
            case this.Trajectory.line:
                return this.computePositionLine( dt );
            case this.Trajectory.parabolic:
                return this.computePositionParabolic( dt );
            default:
                EU.assert( 0 );
        }
        return this.getPosition();
    },
    computePositionLine: function( dt )
    {
        var p0 = this.getPosition();
        var p1 = this._targetPoint;
        var distance = p0.getDistance( p1 );
        var velocity = this.getMover().getVelocity();
        var t = dt*velocity;
        t = Math.min( t, (p1 - p0).getLength() );
        var p = p0 + (p1 - p0).getNormalized() * t;
    
    
        if( EU.Support.checkRadiusByEllipse( p, this._targetPoint, this.getRadius() ) )
            this.push_event( event_die );
    
        return p;
    },
    computePositionParabolic: function( dt )
    {
        /* t = [0;1] */
        var parabolic = function( t ) { return -(2 * t - 1)*(2 * t - 1) + 1; };
        var line = function(t ) { return this._startPoint + (this._targetPoint - this._startPoint) * t; };
    
        this._parabolicParams.timer += dt;
        var t = this._parabolicParams.timer / this._parabolicParams.duration; // 1 - duration;
        t = Math.min( t, 1.0);
        var pos = line( t );
        pos.y += parabolic( t ) * this._parabolicParams.H;
    
        if( t >= 0.999 )
        this.push_event( event_die );
    
        return pos;
    }

});
